### PR TITLE
Snow 1038305 abstraction layer for file reading and writing

### PR DESCRIPTION
**Pull request by sfc-gh-pczajka**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Add SecurePath class which is intended to be used as a default way for handling files by snowcli. Snowcli code updates will happen in next PRs.

### CODE REVIEW

File: src/snowflake/cli/api/exceptions.py [hunk 1]
```diff
@@ -94,3 +95,15 @@ def __init__(self, object_name: str):
...
+class FileTooLargeError(ClickException):
+    def __init__(self, path: Path, size_limit_in_kb: int):
+        super().__init__(
+            f"File {path} is too large (size limit: {size_limit_in_kb} KB)"
+        )
+
+
+class DirectoryIsNotEmptyError(ClickException):
+    def __init__(self, path: Path):
+        super().__init__(f"Directory '{path}' is empty")
```
[the remark this answers is not in the document]
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/exceptions.py:109]: fixed!

File: src/snowflake/cli/api/exceptions.py [hunk 2]
```diff
@@ -94,3 +95,15 @@ def __init__(self, object_name: str):
         super().__init__(
             f"{object_name} defined with secrets but without external integration."
         )
+
+
+class FileTooLargeError(ClickException):
+    def __init__(self, path: Path, size_limit_in_kb: int):
+        super().__init__(
+            f"File {path} is too large (size limit: {size_limit_in_kb} KB)"
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/api/exceptions.py:103]: Did we decide the limit? I think KB is too low, we should have MB.

One important thing, this value should be able to be overwritten by value from `config.toml`. We should not block people to read larger files. WDYT?

File: src/snowflake/cli/api/secure_path.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            if destination.is_dir():
+                destination = destination / self._path.name
+            if destination.exists():
+                raise FileExistsError(destination)
+
+        def _recursive_copy(src: Path, dest: Path):
+            if src.is_file():
+                shutil.copyfile(src, dest)
+            if src.is_dir():
+                dest.mkdir(mode=0o700)
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/api/secure_path.py:161]: not 600?

File: src/snowflake/cli/api/secure_path.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                for child in src.iterdir():
+                    _recursive_copy(child, dest / child.name)
+
+        log.info("Copying %s into %s", self._path, destination)
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/api/secure_path.py:165]: I think we should list all files

File: tests/api/test_secure_path.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+def test_open_write(temp_dir, save_logs):
+    path = SecurePath(temp_dir) / "file.txt"
+    with path.open("w") as fd:
+        # permissions are limited on freshly-created file
+        assert_file_permissions_are_strict(path.path)
+        fd.write("Merlin")
+    assert_file_permissions_are_strict(path.path)
+    logs = _read_logs(save_logs)
+    assert "INFO [snowflake.cli.api.secure_path] Opening file" in logs
+    assert "INFO [snowflake.cli.api.secure_path] Closing file" in logs
```
Comment (sfc-gh-astus) [inline tests/api/test_secure_path.py:75]: Log with file extension would be better

File: src/snowflake/cli/api/secure_path.py (hunk 3 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/secure_path.py:161]: "exec" for directory means "can access elements inside directory" (ls, cd), so user should have it enabled
https://unix.stackexchange.com/questions/21251/execute-vs-read-bit-how-do-directory-permissions-in-linux-work

File: tests/api/test_secure_path.py (hunk 5 above)
Comment (sfc-gh-pczajka) [reply tests/api/test_secure_path.py:75]: Currently whole path is logged:
```
2024-02-09 15:01:13 INFO [snowflake.cli.api.secure_path] Opening file /var/folders/nl/7hy4ty5j7js_84sm1q8zl0y40000gn/T/tmpjgo6tdf2/file.txt in mode 'w'
2024-02-09 15:01:13 INFO [snowflake.cli.api.secure_path] Closing file /var/folders/nl/7hy4ty5j7js_84sm1q8zl0y40000gn/T/tmpjgo6tdf2/file.txt
```

File: tests/api/test_secure_path.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    assert logs.count("file1.txt") == 1 and logs.count("file2.txt") == 2
+
+
+def test_mkdir(temp_dir, save_logs):
+    dir1 = SecurePath(temp_dir) / "dir1"
+    dir2 = SecurePath(temp_dir) / "dir2" / "a" / "b" / "c" / "d"
+
+    dir1.mkdir()
+    _assert_count_current_logs(
+        save_logs, "INFO [snowflake.cli.api.secure_path] Creating directory", 1
```
Comment (sfc-gh-astus) [inline tests/api/test_secure_path.py:162]: We should add `Crating directory /a/b/c/d`

File: src/snowflake/cli/api/secure_path.py (hunk 4 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/secure_path.py]: Done

File: src/snowflake/cli/api/exceptions.py (hunk 2 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/exceptions.py:103]: 1. Done
2. Discussed on daily

**Review by sfc-gh-astus — APPROVED** [review]

**Pull request merged**